---
title: "Bulk RNA-seq Postalignment Pipeline"
author: "Victor Enrique Goitea"
date: "2024-03-05"
output: 
  github_document:
    toc: yes
---

# Overview
This script is designed to process BAM files generated from sequencing data. It performs various quality control (QC) analyses and data preprocessing steps to prepare the data for downstream analysis (read counting, bigWig coverage files). The script utilizes several bioinformatics tools and workflows.

# Prerequisites
* SLURM scheduler.  
*	Modules: SAMtools, Picard tools, Bedtools, Qualimap, Preseq, RNA-SeQC, RSeQC, Subread (FeatureCounts), Homer, Deeptools.  
*	Reference genomes (provided in /work/References).  

# Usage

1. Modify SLURM Parameters (Optional): 
    Open the script (**pe_postalign_RNA-seq_multigenome.sh**) and modify SLURM parameters at the beginning of the file, such as account, output file, email notifications, nodes, memory, CPU cores, and runtime. Alternatively, you can modify these parameters on-the-fly when executing the script.

2. On UCloud, start a **Terminal Ubuntu** run: 
   * Enable **Slurm cluster** 
   * To process several samples consider requesting nodes > 1
   * Set the modules path to
    **FGM \> Utilities \> App \> easybuild**

    ![](../Img/terminal_slurm.png)

   * Include the References folder
  **FGM \> References \> References**

    ![](../Img/terminal_folders.png)

   * Include your Scripts folder and the folder with the bam files and

   * **Notes:** 
     + Match the job CPUs to the amounts requested in the script.
     + Make sure the scripts have executing permission. If not run: `chmod 700 script.sh`
     + If you modify the memory parameter in the script, specify 5-10% less than the memory available in the terminal run.
     + Although it is not necessary to enable **tmux**, it is a good practise to always do it.
     + The script also uses some files in References/Multiqc/ to make some metric files compatible with Multiqc.

3. **Run the Script:** Submit the script to the SLURM cluster:

    ```
    sbatch -J <job_name> path_to/Scripts_folder/pe_postalign_RNA-seq_multigenome.sh -g <mm10|mm39|hg38> <input-bam-file> 
    ```

    **Required Arguments**
    * **-g:** specify the genome to use (mm10, mm39, or hg38).
    * Replace **input-bam-file** with the full path to your input FASTQ file (R1).
    
    For several samples you can use a for loop:
    
    ```
    for i in *.bam; do sbatch -J <job_name> path_to/Scripts_folder/pe_postalign_RNA-seq_multigenome.sh -g <mm10|mm39|hg38> $i; sleep 1; done
    ```

4.	**Monitor Job:** You can monitor the job using the SLURM commands, such as **squeue**, **scontrol show job <job-id>**, and check the log files generated.

# Description
This script performs the following main tasks:

1. **Strandedness determination:** based on the results, it automatically proceeds accordingly, i.e. reverse-stranded, forward-stranded or non-stranded setups. Tools strandness aware include Qualimap, RNA-SeQC and featureCounts.
2. **Extensive quality control:** 
   - **Samtools:** stats, idxstats, flagstat.
   - **Picard:** CollectAlignmentSummaryMetrics, QualityScoreDistribution, MeanQualityByCycle, CollectInsertSizeMetrics, EstimateLibraryComplexity Markduplicates (flag/not remove).
   - **Qualimap:** bamqc, rnaseq.
   - **RNA-seq QC:** QC analyses specific to RNA-seq data
     + **RNA-SeQC**
     + **RSeQC:** read_duplication, read_GC, inner_distance, junction_annotation, junction_saturation, bam_stat, infer_experiment, read_distribution, geneBody_coverage, tin.
   -  **Library Complexity Estimation:** Preseq, Picard and PBC metrics.
3. **Fixing Coodinates:** for example if chromosome nomenclature is "1", convert to "chr1". It keeps only canonical chromosomes including sex and mitochondrial chromosomes.  
4. **Filtering and Sorting:** Filters and sorts the processed BAM files to remove low-quality reads and ensure proper alignment (**samtools -q 30 -F 780**).
5. **Indexing:** indexes the final BAM files.
6. **Bigwig coverage tracks:** generates coverage tracks with bin size 10, RPKM normalized by bamCoverage (Deeptools).
7. **Create Tag Directories:** generates a sample tag directory using Homer for downstream analysis.
8. **Reads Counting:** Count reads mapped to features (e.g., genes) using FeatureCounts (**-p --countReadPairs -t exon**).
9. **GENCODE GTF Version Information:**
   + **Human (GRCh38):** v44
   + **Mouse (GRCm38 - mm10):** M25
   + **Mouse (GRCm39 - mm39):** M33

# Output
The script generates various output files containing QC metrics, alignment statistics, and processed BAM files ready for downstream analysis.
The script store the output files in a directory of basename <input-filename> (Main output directory).

**file-featurecounts.txt:** Table containing the output of FeatureCounts.  
**file-RawCounts.txt:** Two-column table containing GeneID and raw counts.

# Output files reorganization (optional)
After running the script for all samples, each sample will have its own folder with the basename <input-filename> (Main output directory). If you prefer to organize your files by category, you can execute the provided script **reorganize_files_rnaseq.sh** in the terminal.

## Folder Structure:
```
├── FASTQ_Raw
│   ├── *fastq.gz
├── BAM_Raw
│   ├── *bam
│   ├── *bai
├── Align_logs
│   ├── *Log.out
│   ├── *Log.progress.out
│   ├── *SJ.out.tab
│   ├── *Log.final.out
├── BAM_Markdown
│   ├── *dupmark.bam
│   ├── *dupmark.bai
├── BAM_NMSRT (NMSRT = name sorted)
│   │   ├── *nmsrt.bam
├── BAM_NODUP
│   ├── *nodup.bam
│   ├── *nodup.bai
├── BW_COVERAGE
│   ├── *dupmark.bs10.bw
├── TAGDIRECTORIES
├── RAWCOUNTS
│   ├── *featurecounts.txt
│   ├── *RawCounts.txt
├── Metrics
│   ├── QC_FASTQC
│   │   ├── *fastqc.html
│   │   ├── *fastqc.zip
│   ├── QC_fastq_Screen
│   │   ├── *screen.html
│   │   ├── *screen.txt
│   ├── QC_FASTQ_AdapterRemoval
│   │   ├── *settings
│   │   ├── *pair1.truncated.gz
│   │   ├── *pair2.truncated.gz
│   │   ├── *discarded.gz
│   ├── QC_SAMSTATS
│   │   ├── *samstat.qc.txt
│   ├── QC_SAMFLAG
│   │   ├── *flagstat.qc.txt
│   ├── QC_IDXSTAT
│   │   ├── *idxstat.qc.txt
│   ├── QC_PRESEQ
│   │   ├── *libraryComplexity.preseq.qc.tab
│   ├── QC_PBC
│   │   ├── *qc.pbc_mqc.tsv
│   ├── QC_PICARD
│   │   ├── *filt.dup.qc.txt
│   │   ├── *InsertSize.picardMetrics.qc.pdf
│   │   ├── *QByCycle.picardMetrics.qc.txt
│   │   ├── *QScoreDist.picardMetrics.qc.pdf
│   │   ├── *QScoreDist.picardMetrics.qc.txt
│   │   ├── *alignmentSummary.picardMetrics.qc.txt
│   │   ├── *CollectRnaSeqMetrics.picardMetrics.qc.txt
│   │   ├── *CollectRnaSeqMetrics.picardMetrics.qc.pdf
│   ├── QC_Qualimap
│   │   ├── *qc.qualimap
│   │   ├── *qc.qualimap.bamqc
│   ├── QC_Deeptools
│   │   ├── *qc.plotprofile.png
│   │   ├── *qc.plotprofile.txt
│   │   ├── *qc.matrix.scaled.gz
│   │   ├── *qc.estimatereadfiltering.txt
│   │   ├── *qc.coveragePlot.png
│   │   ├── *qc.fingerprints.png
│   │   ├── *qc.fingerprints.txt
│   ├── QC_RNASeQC
│   │   ├── *qc.rnaseqc
│   ├── QC_RSeQC
│   │   ├── *qc.rseqc.GC_plot.r
│   │   ├── *qc.rseqc.bam_stat.txt
│   │   ├── *qc.rseqc.junction.Interact.bed
│   │   ├── *qc.rseqc.geneBodyCoverage.curves.pdf
│   │   ├── *qc.rseqc.junctionSaturation_plot.r
│   │   ├── *.summary.txt
│   │   ├── *qc.rseqc.inner_distance_freq.txt
│   │   ├── *qc.rseqc.read_distribution.txt
│   │   ├── *qc.rseqc.geneBodyCoverage.r
│   │   ├── *qc.rseqc.junctionSaturation_plot.pdf
│   │   ├── *tin.xls
│   │   ├── *qc.rseqc.inner_distance_plot.r
│   │   ├── *qc.rseqc.inner_distance.txt
│   │   ├── *qc.rseqc.tin.summary.txt
│   │   ├── *qc.rseqc.junction.xls
│   │   ├── *qc.rseqc.seq.DupRate.xls
│   │   ├── *qc.rseqc.junction_plot.r
│   │   ├── *qc.rseqc.infer_experiment.txt
│   │   ├── *qc.rseqc.GC_plot.pdf
│   │   ├── *qc.rseqc.junction.bed
│   │   ├── *qc.rseqc.splice_events.pdf
│   │   ├── *qc.rseqc.DupRate_plot.pdf
│   │   ├── *qc.rseqc.splice_junction.pdf
│   │   ├── *qc.rseqc.inner_distance_plot.pdf
│   │   ├── *qc.rseqc.geneBodyCoverage.txt
│   │   ├── *qc.rseqc.GC.xls
│   │   ├── *qc.rseqc.pos.DupRate.xls
│   │   ├── *qc.rseqc.DupRate_plot.r
```
## Creating a Matrix of RawCounts:
If you wish to create a Matrix of RawCounts for all samples, you can follow these steps in the RAW_COUNTS folder:

1. **Create Header:**
```
echo "Geneid" > header.txt
for f in ./*RawCounts.txt; do basename $f .RawCounts.txt; done | cut -c -4 >> header.txt
cat header.txt | paste -s - > Matrix_Raw_counts.txt
```
2. **Combine RawCounts:**
```
paste ./*RawCounts.txt | tail -n +2 | awk 'BEGIN{FS="\t"; OFS=FS}{for (i=3;i<=NF;i+=2) $i=""} 1' | tr -s '\t''\t' >> Matrix_Raw_counts.txt
```
This will create a Matrix_Raw_counts.txt file containing the combined raw counts data from all samples.

# Create a Multiqc report:
If you wish to create a report of the collected metrics, run the following in a ubuntu-terminal job with modules:

```
# load multiQC
module load MultiQC

# Run multiqc in the directory with all the analysis folders:
multiqc -c /work/Refereneces/Multiqc/multiqc_config_preseq_human.yaml ./  
Note: the yaml config file (-c) is optional and it is design to adjust the genome coverage scale in a plot from Preseq. In case of a study in mouse, there is a mouse version of the yaml file in the same directory.
```

**Notes:**  
   - Ensure that the necessary modules are available on your cluster.  
   - The script includes Slurm directives to specify resource requirements. Review and customize the script based on your specific requirements.  
   - For additional information on individual tools and parameters, refer to their official documentation.
  